File: bumpebot_planning/bumpebot_planning/dijkstra_planner.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid, Path
from rclpy.qos import QoSProfile, DurabilityPolicy
from geometry_msgs.msg import PoseStamped, Pose
import rclpy.time
from tf2_ros import TransformListener, Buffer, LookupException

class Dijkstra_planner(Node):
    def __init__(self):
        super().__init__("dijkstra_planner")

        ''' define the required subs and pubs'''
        # map sub for receiving map published by map publisher, also set the durability qos to Transient local
        self.map_qos = QoSProfile(depth = 10)
        self.map_qos = DurabilityPolicy.TRANSIENT_LOCAL
        self.map_sub = self.create_subscription(OccupancyGrid, "/map", self.map_cb, self.map_qos)

        # pose sub to get the goal pose to be sent into the path planner
        self.pose_sub = self.create_subscription(PoseStamped, "/goal_pose", self.goal_cb, 10)

        # Then we need 2 publishers to publish the path and visited map to keep track of the path followed by the bot
        self.path_pub = self.create_publisher(Path, "/dijkstra/path", 10)
        self.map_pub = self.create_publisher(OccupancyGrid, "/dijkstra/visited_map", 10)

        ''' define the variables required'''
        self.map = None
        self.visited_map = OccupancyGrid()
        # to get tf info about the map and the bot we need to get info about their transforms
        self.tf_buffer = Buffer()
        self.tf_listerner = TransformListener(self.tf_buffer, self)
    
    # map_cb is used to update the visited_map to keep track of the nodes travelled by the bot
    def map_cb(self, map_msg:OccupancyGrid):
        self.map = map_msg
        self.visited_map.header = map_msg.header
        self.visited_map.info = map_msg.info
        self.visited_map.data = [-1] * (map_msg.info.height * map_msg.info.width)

    def goal_cb(self, goal_pose:PoseStamped):
        
        while(map):
            try:
                map_to_base_tf = self.tf_buffer.lookup_transform(self.map.header.frame_id, "base_footprint", rclpy.time.Time())
            except LookupException:
                self.get_logger().error("Could not find transform b/w map and base")
                return
        # now we have the tf from map and base, but we need them in pose to be processed for the pathplanner
        map_to_base_pose = Pose()
        map_to_base_pose.position.x = map_to_base_tf.transform.translation.x
        map_to_base_pose.position.y = map_to_base_tf.transform.translation.y
        map_to_base_pose.orientation = map_to_base_tf.transform.rotation

        path = self.plan(map_to_base_pose, goal_pose)
        self.path_pub.publish(path)

    def plan(self, start, goal):
        pass

def main():
    rclpy.init()
    node = Dijkstra_planner()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: bumpebot_planning/src/dijkstra_planner.cpp
#include "bumpebot_planning/dijkstra_planner.hpp"
#include <rmw/qos_profiles.h>

#include <vector>

namespace bumperbot_planning
{
    Dijkistra_planner::Dijkistra_planner() : Node("dijkstra_planner")
    {
        rclcpp::QoS map_qos(10);
        map_qos.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
        map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>
                    ("/map", map_qos, std::bind(&Dijkistra_planner::map_cb, this, std::placeholders::_1));
        pose_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>
                    ("/goal_pose",10, std::bind(&Dijkistra_planner::goal_cb,this,std::placeholders::_1));

        path_pub_ = create_publisher<nav_msgs::msg::Path>("/dijkstra/path",10);
        map_pub_ = create_publisher<nav_msgs::msg::OccupancyGrid>("/dijkstra/visited_map",10);

        tf_buffer = std::make_unique<tf2_ros::Buffer>(get_clock());
        tf_listener = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);

      
        }
        void Dijkistra_planner::map_cb(const nav_msgs::msg::OccupancyGrid::SharedPtr map){
            map_ = map;
            visited_map.header.frame_id = map_->header.frame_id;
            visited_map.info = map->info;
            visited_map.data = std::vector<int8_t>(map_->info.height * map->info.width,-1 ); 

        }
        void Dijkistra_planner::goal_cb(const geometry_msgs::msg::PoseStamped::SharedPtr goal_pose){
            if(!map_){
                RCLCPP_ERROR(get_logger(),("No map received"));
                return;
            }
            visited_map.data = std::vector<int8_t>(map_->info.height * map_->info.width, -1);
            geometry_msgs::msg::TransformStamped map_to_base_tf ;
            try{
                map_to_base_tf = tf_buffer->lookupTransform(map_->header.frame_id, "base_footprint", tf2::TimePointZero);
            }catch(const tf2::TransformException & ex){
                RCLCPP_WARN(get_logger(),"No transform fowund bet the map and the bot");
            }
            geometry_msgs::msg::Pose map_to_base_pose;
            map_to_base_pose.position.x = map_to_base_tf.transform.translation.x;
            map_to_base_pose.position.y = map_to_base_tf.transform.translation.y;
            map_to_base_pose.orientation = map_to_base_tf.transform.rotation;

            auto path_ = plan(map_to_base_pose, goal_pose->pose);
            path_pub_->publish(path_);
        }
        nav_msgs::msg::Path Dijkistra_planner::plan(const geometry_msgs::msg::Pose & start, const geometry_msgs::msg::Pose & goal){
            
        }

        
    } 

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<bumperbot_planning::Dijkistra_planner>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
